### ISSUE TITLE
String parse error with coverage build

### ISSUE THREAD

**Issue by gedaiu** [dlang-community/libdparse#216]
There is something weird about the lexer. This issue is happening on the new compilers DMD 2.78.1 and up. The number of tokens is different when running the following code with coverage:

```d
import std.stdio;
import std.conv;
import std.array;

import dparse.ast;
import dparse.lexer;
import dparse.parser;

void parseFile()
{
  auto f = File("source/app.d");
  immutable auto fileSize = f.size();
  ubyte[] fileBytes = new ubyte[](fileSize.to!size_t);

  if (f.rawRead(fileBytes).length != fileSize)
  {
    return;
  }

  StringCache cache = StringCache(StringCache.defaultBucketCount);

  LexerConfig config;
  config.stringBehavior = StringBehavior.source;
  config.fileName = "app.d";
  config.commentBehavior = CommentBehavior.intern;

  auto lexer = DLexer(fileBytes, config, &cache);

  writeln("====>>", lexer.array.length);
}

void main()
{
  parseFile;
}

unittest
{
  parseFile;
}
```

``` zsh
➜  dparse-error dub test      
No source files found in configuration 'library'. Falling back to "dub -b unittest".
Performing "unittest" build using /usr/bin/dmd for x86_64.
stdx-allocator 2.77.0: building configuration "library"...
libdparse 0.8.0-alpha.5: building configuration "library"...
dparse-error ~master: building configuration "application"...
Linking...
Running ./dparse-error 
====>>229
====>>229
➜  dparse-error dub test --coverage
No source files found in configuration 'library'. Falling back to "dub -b unittest".
Performing "unittest-cov" build using /usr/bin/dmd for x86_64.
stdx-allocator 2.77.0: building configuration "library"...
libdparse 0.8.0-alpha.5: building configuration "library"...
dparse-error ~master: building configuration "application"...
Linking...

Running ./dparse-error 
====>>59
====>>59

```

The tokens without coverage build: 
``` d
["import()", "whitespace( )", "identifier(std)", ".()", "identifier(stdio)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(conv)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(array)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(algorithm)", ";()", "whitespace(\n\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(ast)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(lexer)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(parser)", ";()", "whitespace(\n\n)", "void()", "whitespace( )", "identifier(parseFile)", "(()", ")()", "whitespace(\n)", "{()", "whitespace(\n  )", "auto()", "whitespace( )", "identifier(f)", "whitespace( )", "=()", "whitespace( )", "identifier(File)", "(()", "stringLiteral(\"source/app.d\")", ")()", ";()", "whitespace(\n  )", "immutable()", "whitespace( )", "auto()", "whitespace( )", "identifier(fileSize)", "whitespace( )", "=()", "whitespace( )", "identifier(f)", ".()", "identifier(size)", "(()", ")()", ";()", "whitespace(\n  )", "ubyte()", "[()", "]()", "whitespace( )", "identifier(fileBytes)", "whitespace( )", "=()", "whitespace( )", "new()", "whitespace( )", "ubyte()", "[()", "]()", "(()", "identifier(fileSize)", ".()", "identifier(to)", "!()", "identifier(size_t)", ")()", ";()", "whitespace(\n\n  )", "if()", "whitespace( )", "(()", "identifier(f)", ".()", "identifier(rawRead)", "(()", "identifier(fileBytes)", ")()", ".()", "identifier(length)", "whitespace( )", "!=()", "whitespace( )", "identifier(fileSize)", ")()", "whitespace(\n  )", "{()", "whitespace(\n    )", "return()", ";()", "whitespace(\n  )", "}()", "whitespace(\n\n  )", "identifier(StringCache)", "whitespace( )", "identifier(cache)", "whitespace( )", "=()", "whitespace( )", "identifier(StringCache)", "(()", "identifier(StringCache)", ".()", "identifier(defaultBucketCount)", ")()", ";()", "whitespace(\n\n  )", "identifier(LexerConfig)", "whitespace( )", "identifier(config)", ";()", "whitespace(\n  )", "identifier(config)", ".()", "identifier(stringBehavior)", "whitespace( )", "=()", "whitespace( )", "identifier(StringBehavior)", ".()", "identifier(source)", ";()", "whitespace(\n  )", "identifier(config)", ".()", "identifier(fileName)", "whitespace( )", "=()", "whitespace( )", "stringLiteral(\"app.d\")", ";()", "whitespace(\n  )", "identifier(config)", ".()", "identifier(commentBehavior)", "whitespace( )", "=()", "whitespace( )", "identifier(CommentBehavior)", ".()", "identifier(intern)", ";()", "whitespace(\n\n  )", "auto()", "whitespace( )", "identifier(lexer)", "whitespace( )", "=()", "whitespace( )", "identifier(DLexer)", "(()", "identifier(fileBytes)", ",()", "whitespace( )", "identifier(config)", ",()", "whitespace( )", "&()", "identifier(cache)", ")()", ";()", "whitespace(\n\n  )", "identifier(writeln)", "(()", "stringLiteral(\"====>>\")", ",()", "whitespace( )", "identifier(lexer)", ".()", "identifier(map)", "!()", "(()", "identifier(a)", "whitespace( )", "=>()", "whitespace( )", "identifier(str)", "(()", "identifier(a)", ".()", "identifier(type)", ")()", "whitespace( )", "~()", "whitespace( )", "stringLiteral(\"(\")", "whitespace( )", "~()", "whitespace( )", "identifier(a)", ".()", "identifier(text)", "whitespace( )", "~()", "whitespace( )", "stringLiteral(\")\")", ")()", ".()", "identifier(array)", ")()", ";()", "whitespace(\n)", "}()", "whitespace(\n\n)", "void()", "whitespace( )", "identifier(main)", "(()", ")()", "whitespace(\n)", "{()", "whitespace(\n  )", "identifier(parseFile)", ";()", "whitespace(\n)", "}()", "whitespace(\n\n)", "unittest()", "whitespace(\n)", "{()", "whitespace(\n  )", "identifier(parseFile)", ";()", "whitespace(\n)", "}()", "whitespace(\n)"]
```

tokens with coverage:
``` d
["import()", "whitespace( )", "identifier(std)", ".()", "identifier(stdio)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(conv)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(array)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(std)", ".()", "identifier(algorithm)", ";()", "whitespace(\n\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(ast)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(lexer)", ";()", "whitespace(\n)", "import()", "whitespace( )", "identifier(dparse)", ".()", "identifier(parser)", ";()", "whitespace(\n\n)", "void()", "whitespace( )", "identifier(parseFile)", "(()", ")()", "whitespace(\n)", "{()", "whitespace(\n  )", "auto()", "whitespace( )", "identifier(f)", "whitespace( )", "=()", "whitespace( )", "identifier(File)", "(()", "!ERROR!()"]
```

**Comment by BBasile** [comment]
Seem like a DMD problem. What is app.d ? Same file as the one containing the code you have paste here ? If so you can use `__FILE_FULL_PATH__`.

**Comment by gedaiu** [comment]
`app.d` is a main file created by `dub init`, and yes, it contains the code from the issue.

**Comment by BBasile** [comment]
It can be simplfied to 

```d
import std.stdio;
import std.file;
import std.array;
import dparse.lexer;

void parseFile()
{
  ubyte[] fileBytes = cast(ubyte[]) read(__FILE_FULL_PATH__);
  StringCache cache = StringCache(StringCache.defaultBucketCount);
  LexerConfig config;
  auto lexer = DLexer(fileBytes, config, &cache);
  writeln("====>>", lexer.array.length);
}

void main()
{
  parseFile;
}
```

and gives 98 vs 123

**Comment by BBasile** [comment]
Remove the ASCII from writeln arrow and same number for all builds. Definitively a DMD bug.

**Comment by BBasile** [comment]
I'll close because the problem is caused by code instrumentation. Don't forget to report to D bugzilla.

**Comment by BBasile** [comment]
Reported myself: https://issues.dlang.org/show_bug.cgi?id=18521

**Comment by BBasile** [comment]
It's will be fixed in 2.079.0, verified to work in 2.079.0-rc.1

**Comment by gedaiu** [comment]
awesome! thanks!
